beacon-chain/db/slasherkv/pruning.go

```diff
@@ -0,0 +1,157 @@
+package slasherkv
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+
+	fssz "github.com/ferranbt/fastssz"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	bolt "go.etcd.io/bbolt"
+)
+
+// PruneProposals prunes all proposal data older than currentEpoch - historyLength in
+// specified epoch increments. BoltDB cannot handle long-running transactions, so we instead
+// use a cursor-based mechanism to prune at pruningEpochIncrements by opening individual bolt
+// transactions for each pruning iteration.
+func (s *Store) PruneProposals(
+	ctx context.Context, currentEpoch, pruningEpochIncrements, historyLength types.Epoch,
+) error {
+	if currentEpoch < historyLength {
+		log.Debugf("Current epoch %d < history length %d, nothing to prune", currentEpoch, historyLength)
+		return nil
+	}
+	// We can prune everything less than the current epoch - history length.
+	endEpoch := currentEpoch - historyLength
+	endPruneSlot, err := helpers.StartSlot(endEpoch)
+	if err != nil {
+		return err
+	}
+	encodedEndSlot := fssz.MarshalUint64([]byte{}, uint64(endPruneSlot))
+
+	// We retrieve the lowest stored slot in the proposals bucket.
+	var lowestSlot types.Slot
+	if err = s.db.View(func(tx *bolt.Tx) error {
+		proposalBkt := tx.Bucket(proposalRecordsBucket)
+		c := proposalBkt.Cursor()
+		k, _ := c.First()
+		lowestSlot = slotFromProposalKey(k)
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// If the lowest slot is greater than or equal to the end pruning slot,
+	// there is nothing to prune, so we return early.
+	if lowestSlot >= endPruneSlot {
+		return nil
+	}
+
+	// We prune in increments of `pruningEpochIncrements` at a time to prevent
+	// a long-running bolt transaction which overwhelms CPU and memory.
+	slotCursor := lowestSlot
+	var encodedSlotCursor []byte
+
+	// While we still have epochs to prune based on a cursor, we continue the pruning process.
+	for epochAtCursor := helpers.SlotToEpoch(slotCursor); endEpoch-epochAtCursor > 0; {
```

Comment:
It seems we can get rid of `helpers.SlotToEpoch()` calculation (albeit it doesn't win us much but still). We can calculate epoch once before loop, and then inside the loop we are already calculating the epoch (to see if epoch under cursor is changed), so we can reuse that value and simplify the loop to
```golang
epochAtCursor := helpers.SlotToEpoch(slotCursor)
for  endEpoch-epochAtCursor > 0; {
// ..skip
  slot := slotFromProposalKey(k)
  epoch := helpers.SlotToEpoch(slot) // we can just rename it to `epochAtCursor`, and it will work.
  slotCursor = slot
}
```

Reply:
While on it, can there be a situation (or data saved in database is such that), `epochAtCursor` doesn't grow? This will mean we are in infinite loop. If we have that "necessary epoch's slots will be definitely in DB" invariant guaranteed to us, then no problems, of course.